src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -107,36 +113,65 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
+
+        public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool includeTransitivePackages, bool includeTransitiveOrigins, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            (PackageSpec packageSpec, string assetsFilePath) = await GetCurrentPackageSpecAndAssetsFilePathSafeAsync(token);
+
             if (packageSpec == null) // null means project is not nominated
             {
                 IsInstalledAndTransitiveComputationNeeded = true;
 
-                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
+                return EmptyProjectPackages;
             }
 
-            IReadOnlyList<LockFileTarget> targetsList = null;
+            IList<LockFileTarget> targetsList = null;
             T installedPackages;
-            T transitivePackages;
+            T transitivePackages = default;
             if (IsInstalledAndTransitiveComputationNeeded)
             {
                 // clear the transitive packages cache, since we don't know when a dependency has been removed
                 installedPackages = new T();
-                transitivePackages = new T();
-                targetsList = (await GetTargetsListAsync(assetsPath, token))?.ToList();
+                if (includeTransitivePackages)
+                {
+                    transitivePackages = new T();
+                }
+                targetsList = await GetTargetsListAsync(assetsFilePath, token);
             }
             else
             {
-                if (InstalledPackages == null || TransitivePackages == null)
+                if (InstalledPackages == null)
                 {
                     installedPackages = new T();
-                    transitivePackages = new T();
                 }
                 else
                 {
                     // Make a copy of the caches to prevent concurrency issues.
                     lock (_installedAndTransitivePackagesLock)
                     {
-                        (installedPackages, transitivePackages) = GetInstalledAndTransitivePackagesCacheCopy();
+                        installedPackages = GetCollectionCopy(InstalledPackages);
+                    }
+                }
+
+                if (includeTransitivePackages)
+                {
+                    if (TransitivePackages == null)
+                    {
+                        transitivePackages = new T();
+                    }
+                    else
+                    {
+                        // Make a copy of the caches to prevent concurrency issues.
+                        lock (_installedAndTransitivePackagesLock)
+                        {
+                            transitivePackages = GetCollectionCopy(TransitivePackages);
```

Comment:
Where is the actual new calculation is happening for `transitivePackages`? I didn't visit this code path since last code review. Is it happening inside `ResolvedTransitivePackagesList`?

Reply:
Yes, that's correct.